test-suite/test-squid-conf.sh

```diff
@@ -0,0 +1,74 @@
+#!/bin/sh
+#
+## Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+# Orchestrates a "squid -k parse ..." test of a single Squid configuration
+# file (with an optional .instructions file containing testing directions).
+# Usage: test-squid-conf.sh <top_builddir> <squid.conf>
+
+top_builddir=$1
+configFile=$2
+
+instructionsFile="$configFile.instructions"
```

Comment:
Why not add instructions as a comment to the top of the config file itself? save on all the special case handling for .instructions files.

Reply:
> Why not add instructions as a comment to the top of the config file itself?

I started with that design but eventually rejected it after realizing that combining test metadata (i.e. configuration testing instructions) and data (i.e. the configuration itself) creates more problems than it solves:

* The combination requires special syntax that would allow the testing script to separate metadata from data. There are many ways to do that, with various pros and cons, but even the best segregation approaches require _additional_ parsing/handling in the testing script (that we want to keep simple but which language is not designed for parsing), additional documentation/training, and are still likely to trigger debates about their specific merits (e.g., "why did not you use a special comment symbol/prefix instead of a special section separator?"). It is overall easier/cheaper to separate using files.
* The combination should result in creation of temporary configuration files (the ones without metadata). Those files require _additional_ cleanup code and complicates failure triage/reproduction. [WYSIWYG](https://en.wikipedia.org/wiki/WYSIWYG) is an overall better approach when it comes to such basic test data.
* Combining different syntax, different formatting, different spelling rules, etc. in one file often complicates human editing and automated formatting/validation.
* Restricting test cases to `*.conf` files is arguably an _improvement_ in itself -- the existing `*` approach is arguably too broad/rigid/risky.
* The additional work required to add proper Makefile handling for `.instructions` files is essentially a one-time investment while the extra work caused by mixing metadata and data is likely to be ongoing.

> save on all the special case handling for .instructions files.

The overall net difference would be a loss IMO.

Reply:
Okay, thank you for the details.